### PR TITLE
040 add login system UI

### PR DESCRIPTION
**Pull request by TaylorJCrawford**

### CODE REVIEW

**Review by ghost — APPROVED** [review]

File: app.ts [hunk 1]
```diff
@@ -1,13 +1,40 @@
...
 
+app.set('views', './views');
 app.set('view engine', 'ejs');
 
 app.use('/public', express.static(path.join(__dirname, '/public')));
 
+const bodyParser = require('body-parser');
+app.use(bodyParser.urlencoded({ extended: true }));
+
+app.use(session({ secret: 'NOT HARDCODED SECRET', cookie: {maxAge: 3600000} }));
```
Comment (glenn-hamilton) [inline app.ts]: `NOT HARDCODED SECRET` 😞

File: controllers/AuthController.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            let data: Login = req.body;
+            let authService = new AuthService();
+
+            let tokenRes: String = await authService.loginInUser(data);
+            if (tokenRes == null) {
+                throw new Error;
+            }
+            req.session.token = tokenRes;
+            res.redirect('/');
+        } catch (e) {
```
Comment (glenn-hamilton) [inline controllers/AuthController.ts:20]: Helpfully when `axios` throws an error in this method it leaks the username and password that the user attempted to use

File: test/unit/services/AuthServiceTest.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+};
+
+describe('AuthService', function () {
+    it('Login user, should return token, when valid login details are sent.', async () => {
+        let fakeToken: String = 'Valid Token';
+        mock.onPost('/api/auth/login').reply(200, fakeToken);
+
+        let token = await authService.loginInUser(userAccount);
+
+        expect(token).to.be.an('String');
```
Comment (glenn-hamilton) [inline test/unit/services/AuthServiceTest.ts]: Would this not be better to check that you return the actual value you are expecting, not the type of what is returned?

File: views/pages/home.ejs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<%- include('../templates/components/notification-bar'); %>
+<%- include('../templates/components/common-header'); %>
+
+<body>
+
+    <div class="main-container">
+        <div class="main-content center">
+            Home Page - You have been login.
```
Comment (glenn-hamilton) [inline views/pages/home.ejs]: `Home Page - You have been login.` -> `Home Page - You have been logged in.`

File: views/templates/components/nav.ejs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    </nav>
+</div>
+
+<div id="navItemsLower">
+    <div style="background-color: rgb(255, 255, 255); padding-left: 10vw;">
+        <a class="nav-link navLinkItemCustom" href="/" id="nav-cust-link">Home</a>
+    </div>
+</div>
+
+<script>
```
Comment (glenn-hamilton) [inline views/templates/components/nav.ejs]: I'd rather there not be `<script>` or `<noscript>` tags inside of the `HTML` files, other than linking to other `js` files

(I'd rather there not be any JS tbh)

File: app.ts (hunk 1 above)
Comment (TaylorJCrawford) [reply app.ts:15]: Done

File: controllers/AuthController.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+            let payload = {pageTitle: 'Failed Login', errormessage: 'Login Failed',
+                token: req.session.token};
+            res.render('pages/login', payload);
+        }
+    });
+
+    app.get('/logout',async (req: Request, res: Response) => {
```
Comment (glenn-hamilton) [inline controllers/AuthController.ts]: Logout should be a `post` as it's a destructive action

File: views/templates/components/nav.ejs (hunk 5 above)
Comment (TaylorJCrawford) [reply views/templates/components/nav.ejs]: Fixed! - Was only temp until branding was in.

File: services/AuthService.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import axiosInstance from '../config';
+import { Login } from '../models/Auth';
+
+export default class AuthService {
+
+    async loginInUser(login: Login): Promise<string> {
+
+        // const response = await axiosInstance.post('/api/auth/login', login);
```
Comment (glenn-hamilton) [inline services/AuthService.ts]: Don't need the commented out code

File: test/unit/services/JobServiceTest.ts [hunk 8]
```diff
@@ -1,13 +1,14 @@
...
 let chai = require('chai');
 const expect = chai.expect;
-import { describe, it } from 'mocha';
-import JobService from '../../services/JobService';
-import axiosInstance from '../../config';
-import { Job } from '../../models/Job';
-let mock = new MockAdapter(axiosInstance);
 
 let jobService = new JobService();
+let mock: any;
```
Comment (glenn-hamilton) [inline test/unit/services/JobServiceTest.ts]: Why is `mock` set to `any`?

File: services/AuthService.ts (hunk 7 above)
Comment (TaylorJCrawford) [reply services/AuthService.ts]: Forgot about this. Testing Something.

File: test/unit/services/JobServiceTest.ts (hunk 8 above)
Comment (TaylorJCrawford) [reply test/unit/services/JobServiceTest.ts]: Playing around with object, forgot to change this to type of mock adapter. Thank you! :)

File: app.ts [hunk 9]
```diff
@@ -1,13 +1,40 @@
...
 
+app.set('views', './views');
 app.set('view engine', 'ejs');
 
 app.use('/public', express.static(path.join(__dirname, '/public')));
 
+const bodyParser = require('body-parser');
+app.use(bodyParser.urlencoded({ extended: true }));
+
+app.use(session({ secret: process.env.SESSION_SECRET_UI, cookie: {maxAge: 3600000} }));
```
Comment (glenn-hamilton) [inline app.ts:17]: `SESSION_SECRET_UI` needs added to the GitHub actions and `dockerfile`

File: test/unit/services/AuthServiceTest.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+import axiosInstance from '../../../config';
+import AuthService from '../../../services/AuthService';
+import { Login } from '../../../models/Auth';
+
+let MockAdapter = require('axios-mock-adapter');
+let chai = require('chai');
+const expect = chai.expect;
+
+let authService = new AuthService();
+let mock: any;
```
Comment (glenn-hamilton) [inline test/unit/services/AuthServiceTest.ts]: Is there a way to change this so it isn't `any`?

**Review by glenn-hamilton — APPROVED** [review]

File: test/unit/services/AuthServiceTest.ts (hunk 10 above)
Comment (TaylorJCrawford) [reply test/unit/services/AuthServiceTest.ts]: Yip, sorry forgot to update that one. Thank you.

**Pull request merged**